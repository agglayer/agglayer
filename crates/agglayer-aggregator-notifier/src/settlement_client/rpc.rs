use std::{sync::Arc, time::Duration};

use agglayer_certificate_orchestrator::{Error, SettlementClient};
use agglayer_config::outbound::OutboundRpcSettleConfig;
use agglayer_contracts::{rollup::VerifierType, L1TransactionFetcher, RollupContract, Settler};
use agglayer_storage::stores::{
    PendingCertificateReader, PerEpochReader, PerEpochWriter, StateReader, StateWriter,
};
use agglayer_types::{
    CertificateHeader, CertificateId, CertificateIndex, CertificateStatus, EpochNumber,
    ExecutionMode, Proof, SettlementTxHash,
};
use alloy::{
    providers::{PendingTransactionConfig, Provider},
    rpc::types::TransactionReceipt,
};
use arc_swap::ArcSwap;
use pessimistic_proof::{proof::DisplayToHex, PessimisticProofOutput};
use tracing::{debug, error, info, instrument, warn};

const MAX_EPOCH_ASSIGNMENT_RETRIES: usize = 5;

/// Rpc-based settlement client for L1 certificate settlement.
/// Using alloy client to interact with the L1 rollup manager contract.
#[derive(Default, Clone)]
pub struct RpcSettlementClient<StateStore, PendingStore, PerEpochStore, RollupManagerRpc> {
    state_store: Arc<StateStore>,
    pending_store: Arc<PendingStore>,
    config: Arc<OutboundRpcSettleConfig>,
    l1_rpc: Arc<RollupManagerRpc>,
    current_epoch: Arc<ArcSwap<PerEpochStore>>,
}

impl<StateStore, PendingStore, PerEpochStore, RollupManagerRpc>
    RpcSettlementClient<StateStore, PendingStore, PerEpochStore, RollupManagerRpc>
{
    /// Try to create a new rpc-based settlement client
    pub fn new(
        config: Arc<OutboundRpcSettleConfig>,
        state_store: Arc<StateStore>,
        pending_store: Arc<PendingStore>,
        l1_rpc: Arc<RollupManagerRpc>,
        current_epoch: Arc<ArcSwap<PerEpochStore>>,
    ) -> Self {
        Self {
            config,
            l1_rpc,
            state_store,
            pending_store,
            current_epoch,
        }
    }
}

#[async_trait::async_trait]
impl<StateStore, PendingStore, PerEpochStore, RollupManagerRpc> SettlementClient
    for RpcSettlementClient<StateStore, PendingStore, PerEpochStore, RollupManagerRpc>
where
    StateStore: StateReader + StateWriter + 'static,
    PendingStore: PendingCertificateReader + 'static,
    RollupManagerRpc: RollupContract + Settler + L1TransactionFetcher + Send + Sync + 'static,
    PerEpochStore: PerEpochWriter + PerEpochReader + 'static,
{
    type Provider = alloy::providers::RootProvider<alloy::network::Ethereum>;

    #[instrument(skip(self), fields(network_id, settlement_params), level = "debug")]
    async fn submit_certificate_settlement(
        &self,
        certificate_id: CertificateId,
    ) -> Result<SettlementTxHash, Error> {
        // Step 1: Get certificate header and validate
        let (network_id, height) = if let Some(CertificateHeader {
            status,
            network_id,
            height,
            ..
        }) =
            self.state_store.get_certificate_header(&certificate_id)?
        {
            if status == CertificateStatus::Settled {
                error!("Certificate is already settled");
                return Err(Error::InvalidCertificateStatus);
            }
            (network_id, height)
        } else {
            error!("Certificate header not found");
            return Err(Error::NotFoundCertificateHeader);
        };

        // Step 2: Validate epoch assignment
        let dry_current_epoch = self.current_epoch.load();
        match dry_current_epoch.add_certificate(certificate_id, ExecutionMode::DryRun) {
            Err(error) => {
                drop(dry_current_epoch);
                error!(
                    %error,
                    "{}Failed to add the certificate to the current epoch",
                    ExecutionMode::DryRun.prefix(),
                );
                return Err(Error::Storage(error));
            }
            Ok((_epoch_number, _certificate_index)) => {
                drop(dry_current_epoch);
                info!("Certificate passes the epoch dry run");
            }
        }

        // Step 3: Get certificate from pending store
        let certificate =
            if let Some(certificate) = self.pending_store.get_certificate(network_id, height)? {
                certificate
            } else {
                return Err(Error::InternalError(format!(
                    "Unable to find the certificate {certificate_id} in pending store"
                )));
            };

        let network_id = certificate.network_id;
        tracing::Span::current().record("network_id", network_id.to_u32());

        let l1_info_tree_leaf_count = certificate
            .l1_info_tree_leaf_count()
            .unwrap_or_else(|| self.l1_rpc.default_l1_info_tree_entry().0);

        // Step 4: Deserialize and prepare the proof
        let (output, proof) =
            if let Some(Proof::SP1(proof)) = self.pending_store.get_proof(certificate_id)? {
                if let Ok(output) = PessimisticProofOutput::bincode_codec()
                    .deserialize::<PessimisticProofOutput>(proof.public_values.as_slice())
                {
                    (output, proof.bytes())
                } else {
                    return Err(Error::InternalError(
                        "Unable to deserialize the proof output".to_string(),
                    ));
                }
            } else {
                return Err(Error::InternalError(
                    "Unable to find the proof in the pending store".to_string(),
                ));
            };

        // Step 5: Get verifier type and prepare proof
        let verifier_type = self
            .l1_rpc
            .get_verifier_type(network_id.to_u32())
            .await
            .map_err(|_| Error::UnableToGetVerifierType {
                certificate_id,
                network_id,
            })?;

        debug!("Network {network_id} has {verifier_type:?}");

        let proof_with_selector: Vec<u8> = match verifier_type {
            VerifierType::StateTransition => {
                return Err(Error::InternalError(
                    "Unsupported verifier type".to_string(),
                ));
            }
            VerifierType::Pessimistic => proof,
            VerifierType::ALGateway => {
                let mut proof_with_selector =
                    pessimistic_proof::core::PESSIMISTIC_PROOF_PROGRAM_SELECTOR.to_vec();
                proof_with_selector.extend(&proof);
                proof_with_selector
            }
        };

        info!(
            "Initializing the settlement on L1 with public inputs: {}",
            output.display_to_hex()
        );

        // Record the settlement parameters for tracing
        let settlement_params = format!(
            "origin_network: {}, l1_info_tree_leaf_count: {}, new_local_exit_root: 0x{}, \
             new_pessimistic_root: 0x{}, proof_length: {}, custom_chain_data_length: {}",
            output.origin_network.to_u32(),
            l1_info_tree_leaf_count,
            hex::encode(output.new_local_exit_root),
            hex::encode(output.new_pessimistic_root),
            proof_with_selector.len(),
            certificate.custom_chain_data.len()
        );
        tracing::Span::current().record("settlement_params", &settlement_params);

        // Step 6: Call the contract settlement function and get the pending transaction
        let pending_tx = match self
            .l1_rpc
            .verify_pessimistic_trusted_aggregator(
                output.origin_network.to_u32(),
                l1_info_tree_leaf_count,
                *output.new_local_exit_root.as_ref(),
                *output.new_pessimistic_root,
                proof_with_selector.into(),
                certificate.custom_chain_data.into(),
            )
            .await
        {
            Ok(pending_tx) => {
                info!("Certificate settlement transaction submitted");
                pending_tx
            }
            Err(error) => {
                let error_str = RollupManagerRpc::decode_contract_revert(&error)
                    .unwrap_or_else(|| error.to_string());

                error!(
                    error_code = %error,
                    error = error_str,
                    "Failed to settle certificate"
                );

                return Err(Error::SettlementError {
                    certificate_id,
                    error: error_str,
                });
            }
        };

        // Get the transaction hash from the pending transaction
        let tx_hash = *pending_tx.tx_hash();
        info!("Settlement transaction hash: {}", tx_hash);

        Ok(SettlementTxHash::from(tx_hash))
    }

    #[tracing::instrument(skip(self))]
    async fn wait_for_settlement(
        &self,
        settlement_tx_hash: SettlementTxHash,
        certificate_id: CertificateId,
    ) -> Result<(EpochNumber, CertificateIndex), Error> {
        info!(%settlement_tx_hash, "Waiting for settlement of tx {settlement_tx_hash}");

        // Step 1: Wait for transaction receipt with retries
        let receipt = self
            .wait_for_transaction_receipt(settlement_tx_hash, certificate_id)
            .await?;

        // Apply fail points if they are active for integration testing
        #[cfg(feature = "testutils")]
        testutils::inject_settle_certificate_fail_points(certificate_id)?;

        // Step 2: Check transaction status
        if !receipt.status() {
            warn!(%settlement_tx_hash, "Certificate settlement transaction failed to settle");
            return Err(Error::SettlementError {
                certificate_id,
                error: "Settlement transaction failed".to_string(),
            });
        }

        info!(%settlement_tx_hash, "Certificate settlement transaction successfully settled on l1");

        // Step 3: Add certificate to epoch with retries
        let mut max_retries = MAX_EPOCH_ASSIGNMENT_RETRIES;
        let (epoch_number, certificate_index) = loop {
            max_retries -= 1;

            let related_epoch = self.current_epoch.load_full();
            if related_epoch.is_epoch_packed() {
                drop(related_epoch);
                warn!("The epoch is already packed, adding delay and retry the assignment");
                tokio::time::sleep(Duration::from_secs(1)).await;
                continue;
            }

            match related_epoch.add_certificate(certificate_id, ExecutionMode::Default) {
                Err(error) if max_retries == 0 => {
                    let error_msg = format!(
                        "CRITICAL: Failed to add the certificate to the epoch after multiple \
                         retries: {error}"
                    );
                    error!(%error, error_msg);

                    return Err(Error::SettlementError {
                        certificate_id,
                        error: error_msg,
                    });
                }
                Err(error) => {
                    warn!(%error, "Failed to add the certificate to the epoch (retrying)");
                }
                Ok((epoch_number, certificate_index)) => {
                    info!(
                        "Certificate added to epoch {epoch_number} with index {certificate_index}"
                    );
                    break (epoch_number, certificate_index);
                }
            }
        };

        Ok((epoch_number, certificate_index))
    }
}

impl<StateStore, PendingStore, PerEpochStore, RollupManagerRpc>
    RpcSettlementClient<StateStore, PendingStore, PerEpochStore, RollupManagerRpc>
where
    StateStore: StateReader + StateWriter + 'static,
    PendingStore: PendingCertificateReader + 'static,
    RollupManagerRpc: RollupContract + Settler + L1TransactionFetcher + Send + Sync + 'static,
    PerEpochStore: PerEpochWriter + PerEpochReader + 'static,
{
    /// Wait for transaction receipt with configurable retries and intervals
    async fn wait_for_transaction_receipt(
        &self,
        settlement_tx_hash: SettlementTxHash,
        certificate_id: CertificateId,
    ) -> Result<TransactionReceipt, Error> {
        let tx_hash = settlement_tx_hash.into();
        let timeout = self
            .config
            .retry_interval
            .mul_f64(self.config.max_retries as f64);
        let pending_tx_config = PendingTransactionConfig::new(tx_hash)
            .with_required_confirmations(self.config.confirmations as u64)
            .with_timeout(Some(timeout));

        let pending_tx = self
            .l1_rpc
            .get_provider()
            .watch_pending_transaction(pending_tx_config)
            .await
            .map_err(|e| Error::SettlementError {
                certificate_id,
                error: format!("Failed to watch pending settlement transaction: {e}"),
            })?;

        match pending_tx.await {
            Ok(confirmed_tx_hash) => {
                info!(%settlement_tx_hash, "Transaction confirmed, fetching receipt");
                // Now fetch the actual transaction receipt using the confirmed hash
                self.l1_rpc
                    .fetch_transaction_receipt(confirmed_tx_hash)
                    .await
                    .map_err(|error| {
                        error!(
                            ?error,
                            %settlement_tx_hash,
                            "Failed to fetch settlement transaction receipt"
                        );

                        Error::SettlementError {
                            certificate_id,
                            error: format!(
                                "Failed to fetch settlement transaction receipt: {error}"
                            ),
                        }
                    })
            }
            Err(error) => {
                error!(
                    ?error,
                    %settlement_tx_hash,
                    "Failed to wait for the pending settlement transaction confirmation"
                );
                Err(Error::SettlementError {
                    certificate_id,
                    error: error.to_string(),
                })
            }
        }
    }
}

#[cfg(feature = "testutils")]
mod testutils {
    use agglayer_types::CertificateId;
    use tracing::warn;

    use super::Error;

    pub(crate) fn inject_settle_certificate_fail_points(
        certificate_id: CertificateId,
    ) -> Result<(), Error> {
        // Check if fail points are active and log warnings
        if fail::eval(
            "notifier::packer::settle_certificate::receipt_future_ended::status_0",
            |_| true,
        )
        .unwrap_or(false)
        {
            warn!(
                "FAIL POINT ACTIVE: Simulating transaction receipt with status 0 (failed \
                 transaction)"
            );
            return Err(Error::SettlementError {
                certificate_id,
                error: "Settlement transaction failed (simulated via fail point)".to_string(),
            });
        }

        if fail::eval(
            "notifier::packer::settle_certificate::receipt_future_ended::no_receipt",
            |_| true,
        )
        .unwrap_or(false)
        {
            warn!("FAIL POINT ACTIVE: Simulating no receipt found");
            return Err(Error::SettlementError {
                certificate_id,
                error: "No transaction receipt found (simulated via fail point)".to_string(),
            });
        }

        Ok(())
    }
}
