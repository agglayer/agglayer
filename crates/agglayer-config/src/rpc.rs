use std::{net::Ipv4Addr, str::FromStr, time::Duration};

use jsonrpsee::core::TEN_MB_SIZE_BYTES;
use serde::{Deserialize, Deserializer, Serialize};
use serde_with::serde_as;

/// The default port for the local RPC server.
const DEFAULT_PORT: u16 = 9090;

/// The local RPC server configuration.
#[serde_as]
#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub struct RpcConfig {
    /// If the `PORT` environment variable is set, it will take precedence over
    /// the configuration file.
    #[serde(default = "default_port", deserialize_with = "deserialize_port")]
    pub port: u16,
    #[serde(default = "default_host")]
    pub host: Ipv4Addr,

    /// The maximum size of the request body in bytes.
    #[serde(
        skip_serializing_if = "same_as_default_body_size",
        default = "default_body_size"
    )]
    pub max_request_body_size: u32,
    /// The maximum size of the response body in bytes.
    #[serde(
        skip_serializing_if = "same_as_default_body_size",
        default = "default_body_size"
    )]
    pub max_response_body_size: u32,
    /// The maximum number of connections.
    #[serde(
        skip_serializing_if = "same_as_default_max_connections",
        default = "default_max_connections"
    )]
    pub max_connections: u32,
    /// The maximum number of requests in a batch request. If `None`, the
    /// batch request limit is unlimited.
    #[serde(skip_serializing_if = "crate::default")]
    pub batch_request_limit: Option<u32>,
    /// The interval at which to send ping messages
    #[serde(skip)]
    #[serde_as(as = "Option<crate::with::HumanDuration>")]
    pub ping_interval: Option<Duration>,
    /// Timeout for completion of an RPC request to the AggLayer node.
    #[serde_as(as = "crate::with::HumanDuration")]
    #[serde(default = "default_request_timeout")]
    pub request_timeout: Duration,
}

impl Default for RpcConfig {
    fn default() -> Self {
        Self {
            port: default_port(),
            host: default_host(),
            max_request_body_size: default_body_size(),
            max_response_body_size: default_body_size(),
            max_connections: default_max_connections(),
            batch_request_limit: None,
            ping_interval: None,
            request_timeout: default_request_timeout(),
        }
    }
}

/// The default maximum number of connections.
fn default_max_connections() -> u32 {
    100
}

/// The default size of the request and response bodies in bytes.
fn default_body_size() -> u32 {
    TEN_MB_SIZE_BYTES
}

/// The default port for the local RPC server.
/// If the `PORT` environment variable is set, it will take precedence over
fn default_port() -> u16 {
    from_env_or_default("PORT", DEFAULT_PORT)
}

/// The default host for the local RPC server.
const fn default_host() -> Ipv4Addr {
    Ipv4Addr::new(0, 0, 0, 0)
}

/// Default timeout for completion of an RPC request to the AggLayer node.
const fn default_request_timeout() -> Duration {
    Duration::from_secs(180)
}

fn deserialize_port<'de, D>(deserializer: D) -> Result<u16, D::Error>
where
    D: Deserializer<'de>,
{
    let port = u16::deserialize(deserializer)?;

    Ok(from_env_or_default("PORT", port))
}

/// Get an environment variable or a default value if it is not set.
fn from_env_or_default<T: FromStr>(key: &str, default: T) -> T {
    std::env::var(key)
        .ok()
        .and_then(|value| value.parse().ok())
        .unwrap_or(default)
}

fn same_as_default_body_size(size: &u32) -> bool {
    *size == default_body_size()
}

fn same_as_default_max_connections(max_connections: &u32) -> bool {
    *max_connections == default_max_connections()
}
