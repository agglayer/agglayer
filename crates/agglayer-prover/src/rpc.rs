use agglayer_prover_types::{
    default_bincode_options, v1::proof_generation_service_server::ProofGenerationService, Error,
};
use agglayer_telemetry::prover::{
    PROVING_REQUEST_FAILED, PROVING_REQUEST_RECV, PROVING_REQUEST_SUCCEEDED,
};
use bincode::Options;
use pessimistic_proof::{
    local_exit_tree::hasher::Keccak256Hasher, multi_batch_header::MultiBatchHeader,
    LocalNetworkState,
};
use tower::{buffer::Buffer, util::BoxService, Service, ServiceExt};
use tracing::{debug, error, warn};

use crate::executor::{Request, Response};

pub struct ProverRPC {
    executor: Buffer<BoxService<Request, Response, Error>, Request>,
}

impl ProverRPC {
    pub fn new(executor: Buffer<BoxService<Request, Response, Error>, Request>) -> Self {
        Self { executor }
    }
}

#[tonic::async_trait]
impl ProofGenerationService for ProverRPC {
    async fn generate_proof(
        &self,
        request: tonic::Request<agglayer_prover_types::v1::ProofGenerationRequest>,
    ) -> Result<tonic::Response<agglayer_prover_types::v1::ProofGenerationResponse>, tonic::Status>
    {
        let metrics_attrs = &[];
        PROVING_REQUEST_RECV.add(1, metrics_attrs);
        debug!("Got a request from {:?}", request.remote_addr());

        let request_inner = request.into_inner();
        let initial_state: LocalNetworkState = agglayer_prover_types::default_bincode_options()
            .deserialize(&request_inner.initial_state)
            .map_err(|_| tonic::Status::invalid_argument("Unable to deserialize initial state"))?;

        let batch_header: MultiBatchHeader<Keccak256Hasher> =
            agglayer_prover_types::default_bincode_options()
                .deserialize(&request_inner.batch_header)
                .map_err(|_| {
                    tonic::Status::invalid_argument("Unable to deserialize batch header")
                })?;

        let mut executor = self.executor.clone();
        let executor = executor
            .ready()
            .await
            .map_err(|_error| tonic::Status::internal("Unable to get proof executor"))?;

        match executor
            .call(Request {
                initial_state,
                batch_header,
            })
            .await
        {
            Ok(result) => {
                let response = agglayer_prover_types::v1::ProofGenerationResponse {
                    proof: agglayer_prover_types::default_bincode_options()
                        .serialize(&agglayer_types::Proof::SP1(result.proof))
                        .map_err(|_| {
                            tonic::Status::internal("Unable to serialize generated proof")
                        })?,
                };

                PROVING_REQUEST_SUCCEEDED.add(1, metrics_attrs);
                return Ok(tonic::Response::new(response));
            }
            Err(error) => {
                PROVING_REQUEST_FAILED.add(1, metrics_attrs);
                if let Some(error) = error.downcast_ref::<Error>() {
                    error!("Failed to generate proof: {}", error);
                    let response = match error {
                        Error::UnableToExecuteProver => {
                            tonic::Status::internal("Unable to execute prover")
                        }
                        Error::ExecutorFailed(inner_error) => {
                            if let Ok(bytes_error) = default_bincode_options().serialize(&error) {
                                tonic::Status::with_details(
                                    tonic::Code::InvalidArgument,
                                    error.to_string(),
                                    bytes_error.into(),
                                )
                            } else {
                                warn!("Unable to serialize Execution  error: {}", inner_error);
                                tonic::Status::invalid_argument(error.to_string())
                            }
                        }
                        Error::ProofVerificationFailed(inner_error) => {
                            if let Ok(bytes_error) = default_bincode_options().serialize(&error) {
                                tonic::Status::with_details(
                                    tonic::Code::InvalidArgument,
                                    error.to_string(),
                                    bytes_error.into(),
                                )
                            } else {
                                warn!(
                                    "Unable to serialize SP1 verification error: {}",
                                    inner_error
                                );
                                tonic::Status::invalid_argument(error.to_string())
                            }
                        }
                        Error::ProverFailed(_) => tonic::Status::internal(error.to_string()),
                    };

                    return Err(response);
                } else {
                    error!("Failed to generate proof: {:?}", error);

                    return Err(tonic::Status::internal("Failed to generate proof"));
                }
            }
        }
    }
}
