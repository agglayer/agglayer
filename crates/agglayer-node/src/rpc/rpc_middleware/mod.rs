//! RPC middleware layers.

use jsonrpsee::server::middleware::rpc::RpcServiceBuilder;
use tower::layer::util::{Identity, Stack};

mod logging_timeout;

pub use logging_timeout::LoggingTimeoutLayer;

/// The stack of RPC middleware layers.
pub type RpcStack = Stack<LoggingTimeoutLayer, Identity>;

/// Build the RPC middleware stack from config.
pub fn build(config: &agglayer_config::Config) -> RpcServiceBuilder<RpcStack> {
    jsonrpsee::server::middleware::rpc::RpcServiceBuilder::new()
        .layer(LoggingTimeoutLayer::new(config.rpc.request_timeout))
}
