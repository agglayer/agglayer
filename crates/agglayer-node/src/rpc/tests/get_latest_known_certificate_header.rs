use std::{sync::Arc, time::Duration};

use agglayer_config::Config;
use agglayer_storage::{
    storage::{pending_db_cf_definitions, state_db_cf_definitions, DB},
    stores::{pending::PendingStore, state::StateStore, PendingCertificateWriter, StateWriter},
    tests::TempDBDir,
};
use agglayer_types::{Certificate, CertificateHeader, CertificateStatus};
use insta::assert_snapshot;
use jsonrpsee::{core::client::ClientT, rpc_params};
use rstest::*;
use serde_json::json;

use super::TestContext;
use crate::rpc::AgglayerServer as _;

#[rstest]
#[test_log::test(tokio::test)]
async fn returns_the_pending_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);
    let proven_certificate = Certificate::new_for_test(network_id, 1);
    let pending_certificate = Certificate::new_for_test(network_id, 2);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&proven_certificate, CertificateStatus::Proven)
        .expect("unable to insert proven certificate header");
    state_db
        .insert_certificate_header(&pending_certificate, CertificateStatus::Pending)
        .expect("unable to insert pending certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &0,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    pending_db
        .set_latest_proven_certificate_per_network(&network_id, &1, &proven_certificate.hash())
        .expect("unable to set latest proven certificate");

    pending_db
        .insert_pending_certificate(network_id, 2, &pending_certificate)
        .expect("unable to insert pending certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, pending_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Pending);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::pending", json);
}

#[test_log::test(tokio::test)]
async fn returns_the_proven_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);
    let proven_certificate = Certificate::new_for_test(network_id, 1);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&proven_certificate, CertificateStatus::Proven)
        .expect("unable to insert proven certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &settled_certificate.height,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");
    pending_db
        .set_latest_proven_certificate_per_network(&network_id, &1, &proven_certificate.hash())
        .expect("unable to set latest proven certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, proven_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Proven);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::proven", json);
}

#[test_log::test(tokio::test)]
async fn returns_the_settled_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 0);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &settled_certificate.height,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, settled_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Settled);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!(
        "get_latest_known_certificate_header::settled_certificate",
        json
    );
}

#[test_log::test(tokio::test)]
async fn returns_no_certificate_header() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);

    let network_id = 1;

    let context = TestContext::new_with_config(config).await;

    let payload: Option<CertificateHeader> = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert!(payload.is_none());

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!(
        "get_latest_known_certificate_header::no_certificate_header",
        json
    );
}

#[test_log::test(tokio::test)]
async fn returns_the_highest_height() {
    let path = TempDBDir::new();

    let config = Config::new(&path.path);
    let pending_db = Arc::new(
        DB::open_cf(&config.storage.pending_db_path, pending_db_cf_definitions())
            .expect("unable to open pending db"),
    );
    let state_db = Arc::new(
        DB::open_cf(&config.storage.state_db_path, state_db_cf_definitions())
            .expect("unable to open state db"),
    );

    let state_db = StateStore::new(state_db);
    let pending_db = PendingStore::new(pending_db);
    let network_id = 1.into();

    let settled_certificate = Certificate::new_for_test(network_id, 10);
    let pending_certificate = Certificate::new_for_test(network_id, 3);

    state_db
        .insert_certificate_header(&settled_certificate, CertificateStatus::Settled)
        .expect("unable to insert settled certificate header");
    state_db
        .insert_certificate_header(&pending_certificate, CertificateStatus::Pending)
        .expect("unable to insert pending certificate header");

    state_db
        .set_latest_settled_certificate_for_network(
            &network_id,
            &10,
            &settled_certificate.hash(),
            &0,
            &0,
        )
        .expect("unable to set latest settled certificate");

    pending_db
        .insert_pending_certificate(network_id, 3, &pending_certificate)
        .expect("unable to insert pending certificate");

    drop(pending_db);
    drop(state_db);

    let context = TestContext::new_with_config(config).await;

    let payload: CertificateHeader = context
        .client
        .request(
            "interop_getLatestKnownCertificateHeader",
            rpc_params![network_id],
        )
        .await
        .unwrap();

    assert_eq!(payload.certificate_id, settled_certificate.hash());
    assert_eq!(payload.status, CertificateStatus::Settled);
    assert_eq!(payload.height, 10);

    drop(context);

    // Have some delayu to ensure that the server has been stopped
    tokio::time::sleep(Duration::from_millis(100)).await;

    let config = Config::new(&path.path);
    // Restarting the server in raw mode
    let raw_rpc = TestContext::new_raw_rpc_with_config(config).await;
    let rpc = raw_rpc.rpc.into_rpc();
    let payload = json!({
        "jsonrpc": "2.0",
        "method": "interop_getLatestKnownCertificateHeader",
        "params": vec![network_id],
        "id": 0
    });
    let (response, _) = rpc
        .raw_json_request(&serde_json::to_string(&payload).unwrap(), 1)
        .await
        .unwrap();
    let json = serde_json::from_str::<serde_json::Value>(&response).unwrap();
    let json = serde_json::to_string_pretty(&json).unwrap();

    assert_snapshot!("get_latest_known_certificate_header::highest_height", json);
}
