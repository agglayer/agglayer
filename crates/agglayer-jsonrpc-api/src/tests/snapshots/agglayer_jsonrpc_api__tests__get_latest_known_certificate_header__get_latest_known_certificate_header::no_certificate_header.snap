---
source: crates/agglayer-node/src/rpc/tests/get_latest_known_certificate_header.rs
expression: json
---
{
  "id": 0,
  "jsonrpc": "2.0",
  "result": null
}
